File: scripts/dev/journal.py
#!/usr/bin/env python3
"""Read plan/journal/*.md at git HEAD, report problem classes with 2+ rows.

The journal is one file per problem class, each holding a markdown table
whose rows are occurrences.  Recurrence is the row count.  A second row
in a file is the alarm.

    python3 scripts/dev/journal.py
    python3 scripts/dev/journal.py --repo /path/to/checkout

Exit 0 when nothing to report.  Exit 1 on a malformed table or an
unparseable Date.  Exit 2 when the journal cannot be read at all.

Reading HEAD rather than the working tree is deliberate: a working-tree read
counts another session's unlanded rows.  It also means a journal that is not
committed yet is INVISIBLE to the count, so the three outcomes are kept apart --
an empty result is only legitimate when the working tree has no journal
either, and a git failure is an error rather than a silent zero.  A class file
on disk that HEAD does not carry is named on stderr, so the count is never read
as the whole journal.
"""

from __future__ import annotations

import argparse
import re
import subprocess
import sys
from datetime import date
from pathlib import Path

JOURNAL_DIR = "plan/journal"
JOURNAL_HEADER_CELLS = ("date", "spec", "surface", "symptom", "fix")
JOURNAL_SEPARATOR_RE = re.compile(r"^:?-{2,}:?$")

# The single cell a malformed row carries. `journal_row_cells()` returns
# [MALFORMED] rather than None so a caller cannot mistake a broken row for
# prose and skip it (AC-4: a malformed row is named, never skipped).
MALFORMED = "MALFORMED"


class JournalError(RuntimeError):
    """The journal could not be read, so no verdict about it is possible.

    Raised instead of returning an empty result, because "no rows" and "I
    could not look" are different answers and only one of them is green.
    """


def journal_row_cells(line: str) -> list[str] | None:
    """The five cells of a journal table row, or None for non-row lines.

    Returns None for prose and for the table header/separator.  A line that
    IS a row but does not split into exactly five cells returns [MALFORMED],
    and so does a row that lost its leading pipe: inside a table such a line
    is a row, and skipping it silently is what AC-4 forbids.  A journal file
    is repo-authored, so prose in one must not contain a `|`.

    This is the ONE implementation.  `commit_helper.py` and
    `spec-closure-check.py` import it, the way `deferral_orphans.py` imports
    `_deferral_row_cells` from `commit_helper`.
    """
    if "|" not in line:
        return None
    if not line.lstrip().startswith("|"):
        return [MALFORMED]
    fields = line.split("|")
    # A well-formed `| a | b | c | d | e |` splits into 7: leading and
    # trailing empties around the five cells.
    if len(fields) != 7 or fields[0].strip() or fields[-1].strip():
        return [MALFORMED]
    cells = [f.strip() for f in fields[1:6]]
    if tuple(c.lower() for c in cells) == JOURNAL_HEADER_CELLS:
        return None
    if all(JOURNAL_SEPARATOR_RE.match(c) for c in cells):
        return None
    return cells


# A Spec cell names the spec stems a row belongs to, or says the row belongs to
# no spec. `-` is the documented "no spec" (plan/journal/README.md); rows also
# write `none`, add a parenthesised note, and list two stems separated by a
# comma, so all four shapes are read here rather than at each call site.
JOURNAL_SPEC_STEM_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
JOURNAL_SPEC_NONE = ("", "-", "none", "n/a")
_JOURNAL_SPEC_NOTE_RE = re.compile(r"\s*\([^()]*\)\s*$")


def journal_spec_stems(cell: str) -> list[str] | None:
    """The spec stems a Spec cell names: [] for none, None when unreadable.

    A stem is the `plan/spec-<stem>.md` name, and it is a KEY: the gates look
    up `tmp/review/<stem>-<session>.md` under it. So the cell's prose must not
    reach them. `commit_helper.py` took the cell verbatim, and a row saying
    `none (walked into during <spec> closure)` asked the review gate for an
    artifact filed under that sentence. The gate then blocked a commit that
    owed no review, and its error named a path nobody could write.

    Four shapes are read, because the journal already holds all four:

    * `-`, `none`, an empty cell -> `[]`. The row belongs to no spec.
    * `some-spec` -> `["some-spec"]`.
    * `some-spec (measurement only)` -> `["some-spec"]`. A trailing
      parenthesised note is the author writing to the next reader, and it is
      never part of the key.
    * `spec-a, spec-b` -> both. One row can name two specs.

    Anything else returns None, which means UNREADABLE and does not mean "no
    spec". That difference is load-bearing: `commit_helper.journal_row_problems`
    blocks the commit on None, and only that block makes the skip in
    `_journal_added_spec_stems` safe. Read None as `[]` there and a cell nobody
    can parse takes the review gate off the commit that carries the code.
    """
    text = _JOURNAL_SPEC_NOTE_RE.sub("", cell.strip())
    stems: list[str] = []
    for token in text.split(","):
        token = token.strip()
        if token.lower() in JOURNAL_SPEC_NONE:
            continue
        if not JOURNAL_SPEC_STEM_RE.match(token):
            return None
        if token not in stems:
            stems.append(token)
    return stems


def journal_rows(text: str) -> list[list[str]]:
    """Every table row in one journal file's text."""
    rows: list[list[str]] = []
    for line in text.splitlines():
        cells = journal_row_cells(line)
        if cells is not None:
            rows.append(cells)
    return rows


def _git(repo: Path, *args: str) -> subprocess.CompletedProcess[str]:
    try:
        return subprocess.run(
            ["git", *args],
            capture_output=True,
            text=True,
            cwd=repo,
        )
    except OSError as exc:  # git absent, repo path gone
        raise JournalError(f"cannot run git in {repo}: {exc}") from exc


def _class_files_on_disk(repo: Path) -> list[str]:
    """Journal class files in the WORKING TREE, as repo-relative paths."""
    return sorted(
        p.relative_to(repo).as_posix()
        for p in (repo / JOURNAL_DIR).glob("*.md")
        if p.name != "README.md"
    )


def read_journal_at_head(repo: Path) -> tuple[dict[str, list[list[str]]], list[str]]:
    """Read plan/journal/*.md from git HEAD: the rows, and what was NOT read.

    Returns {class_name: rows} from HEAD, plus the class files the working tree
    holds that HEAD does not.  The second half is what keeps the first honest.
    This detector reads HEAD by design -- a working-tree read counts another
    session's unlanded rows -- so an uncommitted class file contributes nothing
    to the count, and the reader is told which files those were.

    Three outcomes, kept distinct:

    * HEAD carries journal CLASS files -> their rows, and the names of any class
      file on disk that HEAD does not carry.  `report()` prints those names and
      still exits 0.  A red here would fire on every session that opens a NEW
      problem class, at the moment `/ze-close` runs the pre-commit gate over the
      row it has just written and not yet committed, and the only way out would
      be to reuse a class that does not fit.  A guard that pressures a writer to
      misfile is worse than the miss it prevents; naming the file removes the
      SILENCE, which is the defect.
    * HEAD carries none AND the working tree holds no class file either -> {}.
      There is no journal, so zero rows is the truth.
    * HEAD carries none while the working tree holds class files -> JournalError.
      Nothing at all was read, so the verdict is vacuous and there is no output
      to carry the caveat.  A git invocation that fails raises for that reason
      too: "I could not look" is not "nothing to report".

    The on-disk comparison keys on `paths`, never on `names`.  `names` counts
    `README.md`, which carries no occurrence and is skipped everywhere else, so
    a HEAD holding the README and no class file made `names` non-empty, took the
    guard's early exit, and reported zero rows over a working tree full of
    classes.  That is the exact vacuous green the guard exists to refuse, and
    the README is committed one ordering step before the first class file.
    """
    listing = _git(repo, "ls-tree", "--name-only", "-r", "HEAD", "--", JOURNAL_DIR)
    if listing.returncode != 0:
        stderr = listing.stderr.strip() or f"exit {listing.returncode}"
        raise JournalError(f"git ls-tree HEAD -- {JOURNAL_DIR} failed: {stderr}")

    names = [n.strip() for n in listing.stdout.splitlines() if n.strip()]
    paths = [n for n in names if n.endswith(".md") and not n.endswith("/README.md")]
    on_disk = _class_files_on_disk(repo)
    if not paths:
        if on_disk:
            raise JournalError(
                f"{JOURNAL_DIR}/ exists in the working tree but HEAD carries no "
                "journal class file.\n"
                "  This detector reads HEAD, so it would count zero rows over an "
                "uncommitted journal\n"
                f"  and report no recurrence at all.  Commit {JOURNAL_DIR}/."
            )
        return {}, []
    unread = [path for path in on_disk if path not in set(paths)]

    result: dict[str, list[list[str]]] = {}
    for path in paths:
        content = _git(repo, "show", f"HEAD:{path}")
        if content.returncode != 0:
            stderr = content.stderr.strip() or f"exit {content.returncode}"
            raise JournalError(f"git show HEAD:{path} failed: {stderr}")
        rows = journal_rows(content.stdout)
        if rows:
            result[path[len(JOURNAL_DIR) + 1 :].removesuffix(".md")] = rows
    return result, unread


def _parse_date(cell: str) -> date | None:
    """The row's Date as a date, or None when the cell is not YYYY-MM-DD."""
    try:
        return date.fromisoformat(cell.strip())
    except (ValueError, AttributeError):
        return None


def report(repo: Path) -> int:
    """Print classes with 2+ rows.  Return non-zero on an unreadable row.

    An unparseable Date is an ERROR, not a footnote: the span is the only
    thing the report says beyond the count, and a class whose dates do not
    parse silently loses it.  A seed of 52 rows carrying `-` in every Date
    cell passed every gate before this returned non-zero.

    A class file on disk that HEAD does not carry is NAMED on stderr and does
    not fail the run.  The count below is over HEAD, so those rows are not in
    it; saying nothing would make the number read as the whole journal
    (`ai/rules/evidence.md`).
    """
    classes, unread = read_journal_at_head(repo)

    problems: list[str] = []
    lines: list[str] = []

    for class_name in sorted(classes):
        rows = classes[class_name]
        rel = f"{JOURNAL_DIR}/{class_name}.md"

        broken = sum(1 for row in rows if row == [MALFORMED])
        if broken:
            problems.append(
                f"MALFORMED: {rel}: {broken} row(s) do not hold the five cells "
                "| Date | Spec | Surface | Symptom | Fix |"
            )
            continue

        dates = [_parse_date(row[0]) for row in rows]
        unparseable = [row[0] for row, d in zip(rows, dates) if d is None]
        if unparseable:
            shown = ", ".join(repr(cell) for cell in unparseable[:3])
            problems.append(
                f"UNPARSEABLE DATE: {rel}: {shown} -- every row needs a "
                "YYYY-MM-DD Date, which is what the span is computed from"
            )
            continue

        if len(rows) < 2:
            continue

        first = min(d for d in dates if d is not None)
        last = max(d for d in dates if d is not None)
        lines.append(
            f"{class_name}: {len(rows)} rows, "
            f"{(last - first).days}d span ({first} .. {last})"
        )

    for line in lines:
        print(line)
    for path in unread:
        print(
            f"NOT AT HEAD: {path} is on disk and not committed, so its rows are "
            "not in the counts above",
            file=sys.stderr,
        )
    if problems:
        for problem in problems:
            print(problem, file=sys.stderr)
        return 1
    return 0


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Report problem journal classes with 2+ rows.",
    )
    parser.add_argument(
        "--repo",
        default=None,
        help="repository root to read (default: this checkout)",
    )
    args = parser.parse_args(argv)

    repo = (
        Path(args.repo).resolve()
        if args.repo
        else Path(__file__).resolve().parent.parent.parent
    )
    try:
        return report(repo)
    except JournalError as exc:
        print(f"journal: {exc}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    sys.exit(main())

File: internal/le/journal/journal.go
// Design: docs/architecture/core-design.md -- native le tool implementation
//
// Package journal reads plan/journal class files from git HEAD. The working
// tree is used only to name class files that the committed count did not read.
// A malformed row or date is a refusal, never an omitted occurrence.
package journal

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
	"unicode"

	"github.com/ze-software/ze/internal/core/textbuf"
	"github.com/ze-software/ze/internal/le/lepath"
)

const (
	area          = "journal"
	journalDir    = "plan/journal"
	journalPrefix = "plan/journal/"
	journalReadme = "README.md"
	markdownExt   = ".md"
)

var journalHeader = [...]string{"date", "spec", "surface", "symptom", "fix"}

type journalRow struct {
	cells     [5]string
	malformed bool
}

type journalClass struct {
	name string
	rows []journalRow
}

type gitAnswer struct {
	stdout   []byte
	stderr   string
	code     int
	startErr error
}

// reportHere runs the report action over this checkout.
func reportHere() (any, int) {
	tree, err := lepath.Root()
	if err != nil {
		var tb textbuf.Buffer
		fmt.Fprintln(os.Stderr, tb.Str("journal: ").Err(err).String()) //nolint:errcheck // CLI output
		return nil, 2
	}

	report, code := Run(tree, os.Stderr)
	if code == 2 {
		return nil, code
	}
	return report, code
}

// Run reads one checkout, writes its warnings and refusals to errOut, and
// returns its structured report and exact gate code.
func Run(tree string, errOut io.Writer) (Report, int) {
	report, err := Check(tree)
	if err != nil {
		var tb textbuf.Buffer
		if _, writeErr := fmt.Fprintln(errOut, tb.Str("journal: ").Err(err).String()); writeErr != nil {
			return Report{}, 2
		}
		return Report{}, 2
	}

	for _, path := range report.Unread {
		var tb textbuf.Buffer
		if _, err := fmt.Fprintln(errOut, tb.Str("NOT AT HEAD: ").Str(path).
			Str(" is on disk and not committed, so its rows are not in the counts above").String()); err != nil {
			return report, 2
		}
	}
	for _, problem := range report.Problems {
		if _, err := fmt.Fprintln(errOut, problem.Text()); err != nil {
			return report, 2
		}
	}
	if len(report.Problems) > 0 {
		return report, 1
	}
	return report, 0
}

// Check reads every journal class at git HEAD and returns the trustworthy
// recurring classes, worktree-only files, and refused classes.
func Check(tree string) (Report, error) {
	paths, err := classFilesAtHead(tree)
	if err != nil {
		return Report{}, err
	}
	onDisk, err := classFilesOnDisk(tree)
	if err != nil {
		return Report{}, err
	}

	if len(paths) == 0 {
		if len(onDisk) > 0 {
			var tb textbuf.Buffer
			return Report{}, errors.New(tb.Str(journalDir).
				Str("/ exists in the working tree but HEAD carries no journal class file.\n").
				Str("  This detector reads HEAD, so it would count zero rows over an uncommitted journal\n").
				Str("  and report no recurrence at all.  Commit ").Str(journalDir).Str("/.").String())
		}
		return Report{}, nil
	}

	report := Report{Unread: unreadPaths(paths, onDisk)}
	classes, err := readClasses(tree, paths)
	if err != nil {
		return Report{}, err
	}
	for _, class := range classes {
		appendClass(&report, class)
	}
	return report, nil
}

func classFilesAtHead(tree string) ([]string, error) {
	answer := runGit(tree, "ls-tree", "--name-only", "-r", "HEAD", "--", journalDir)
	if answer.startErr != nil {
		var tb textbuf.Buffer
		return nil, errors.New(tb.Str("cannot run git in ").Str(tree).Str(": ").Err(answer.startErr).String())
	}
	if answer.code != 0 {
		return nil, gitReadError("git ls-tree HEAD -- plan/journal failed: ", answer)
	}

	var paths []string
	for line := range strings.SplitSeq(string(answer.stdout), "\n") {
		path := strings.TrimSpace(line)
		if path == "" {
			continue
		}
		if !strings.HasSuffix(path, markdownExt) {
			continue
		}
		if strings.HasSuffix(path, "/README.md") {
			continue
		}
		paths = append(paths, path)
	}
	return paths, nil
}

func classFilesOnDisk(tree string) ([]string, error) {
	dir := filepath.Join(tree, filepath.FromSlash(journalDir))
	entries, err := os.ReadDir(dir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	paths := make([]string, 0, len(entries))
	for _, entry := range entries {
		name := entry.Name()
		if name == journalReadme {
			continue
		}
		if !strings.HasSuffix(name, markdownExt) {
			continue
		}
		paths = append(paths, filepath.ToSlash(filepath.Join(journalDir, name)))
	}
	sort.Strings(paths)
	return paths, nil
}

func unreadPaths(head, onDisk []string) []string {
	committed := make(map[string]struct{}, len(head))
	for _, path := range head {
		committed[path] = struct{}{}
	}

	var unread []string
	for _, path := range onDisk {
		if _, ok := committed[path]; ok {
			continue
		}
		unread = append(unread, path)
	}
	return unread
}

func readClasses(tree string, paths []string) ([]journalClass, error) {
	classes := make([]journalClass, 0, len(paths))
	for _, path := range paths {
		answer := runGit(tree, "show", headObject(path))
		if answer.startErr != nil {
			var tb textbuf.Buffer
			return nil, errors.New(tb.Str("cannot run git in ").Str(tree).Str(": ").Err(answer.startErr).String())
		}
		if answer.code != 0 {
			var prefix textbuf.Buffer
			prefix.Str("git show HEAD:").Str(path).Str(" failed: ")
			return nil, gitReadError(prefix.String(), answer)
		}

		rows := journalRows(string(answer.stdout))
		if len(rows) == 0 {
			continue
		}
		name := strings.TrimSuffix(strings.TrimPrefix(path, journalPrefix), markdownExt)
		classes = append(classes, journalClass{name: name, rows: rows})
	}
	sort.Slice(classes, func(left, right int) bool { return classes[left].name < classes[right].name })
	return classes, nil
}

func headObject(path string) string {
	var tb textbuf.Buffer
	return tb.Str("HEAD:").Str(path).String()
}

func runGit(tree string, args ...string) gitAnswer {
	// Git is the gate's data source. The command is bounded by local repository
	// I/O and has no network or lock wait. A timeout could turn a slow disk into
	// a false report that HEAD has no journal.
	cmd := exec.CommandContext(context.Background(), "git", args...) //nolint:gosec // fixed git verbs over a checkout chosen by lepath or a test
	cmd.Dir = tree
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	stdout, err := cmd.Output()
	answer := gitAnswer{stdout: stdout, stderr: strings.TrimSpace(stderr.String())}
	if err == nil {
		return answer
	}

	if exit, ok := errors.AsType[*exec.ExitError](err); ok {
		answer.code = exit.ExitCode()
		return answer
	}
	answer.startErr = err
	return answer
}

func gitReadError(prefix string, answer gitAnswer) error {
	var tb textbuf.Buffer
	tb.Str(prefix)
	if answer.stderr != "" {
		return errors.New(tb.Str(answer.stderr).String())
	}
	return errors.New(tb.Str("exit ").Int(int64(answer.code)).String())
}

func journalRows(text string) []journalRow {
	var rows []journalRow
	for line := range strings.SplitSeq(text, "\n") {
		row, ok := parseJournalRow(line)
		if ok {
			rows = append(rows, row)
		}
	}
	return rows
}

func parseJournalRow(line string) (journalRow, bool) {
	if !strings.ContainsRune(line, '|') {
		return journalRow{}, false
	}
	if !strings.HasPrefix(strings.TrimLeftFunc(line, unicode.IsSpace), "|") {
		return journalRow{malformed: true}, true
	}

	fields := strings.Split(line, "|")
	if len(fields) != 7 {
		return journalRow{malformed: true}, true
	}
	if strings.TrimSpace(fields[0]) != "" {
		return journalRow{malformed: true}, true
	}
	if strings.TrimSpace(fields[6]) != "" {
		return journalRow{malformed: true}, true
	}

	var row journalRow
	for index := range row.cells {
		row.cells[index] = strings.TrimSpace(fields[index+1])
	}
	if isJournalHeader(row.cells) {
		return journalRow{}, false
	}
	if isJournalSeparator(row.cells) {
		return journalRow{}, false
	}
	return row, true
}

func isJournalHeader(cells [5]string) bool {
	for index, cell := range cells {
		if !strings.EqualFold(cell, journalHeader[index]) {
			return false
		}
	}
	return true
}

func isJournalSeparator(cells [5]string) bool {
	for _, cell := range cells {
		if !isSeparatorCell(cell) {
			return false
		}
	}
	return true
}

func isSeparatorCell(cell string) bool {
	cell = strings.TrimSpace(cell)
	cell = strings.TrimPrefix(cell, ":")
	cell = strings.TrimSuffix(cell, ":")
	if len(cell) < 2 {
		return false
	}
	for _, one := range cell {
		if one != '-' {
			return false
		}
	}
	return true
}

func appendClass(report *Report, class journalClass) {
	malformed := 0
	for _, row := range class.rows {
		if row.malformed {
			malformed++
		}
	}
	path := classPath(class.name)
	if malformed > 0 {
		report.Problems = append(report.Problems, Problem{
			Class: class.name,
			Path:  path,
			Kind:  ProblemMalformed,
			Rows:  malformed,
		})
		return
	}

	dates := make([]time.Time, 0, len(class.rows))
	var invalid []string
	for _, row := range class.rows {
		parsed, err := time.Parse(time.DateOnly, row.cells[0])
		if err != nil {
			invalid = append(invalid, row.cells[0])
			continue
		}
		dates = append(dates, parsed)
	}
	if len(invalid) > 0 {
		report.Problems = append(report.Problems, Problem{
			Class:        class.name,
			Path:         path,
			Kind:         ProblemUnparseableDate,
			InvalidDates: invalid[:min(3, len(invalid))],
		})
		return
	}
	if len(class.rows) < 2 {
		return
	}

	first := dates[0]
	last := dates[0]
	for _, parsed := range dates[1:] {
		if parsed.Before(first) {
			first = parsed
		}
		if parsed.After(last) {
			last = parsed
		}
	}
	report.Classes = append(report.Classes, Class{
		Name:     class.name,
		Rows:     len(class.rows),
		SpanDays: int(last.Sub(first).Hours() / 24),
		First:    first.Format(time.DateOnly),
		Last:     last.Format(time.DateOnly),
	})
}

func classPath(name string) string {
	var tb textbuf.Buffer
	return tb.Str(journalPrefix).Str(name).Str(markdownExt).String()
}

// specStems parses the canonical Spec-cell vocabulary. The boolean distinguishes
// an unreadable key from a row that explicitly names no spec.
func specStems(cell string) ([]string, bool) {
	text := strings.TrimSpace(cell)
	if strings.HasSuffix(text, ")") {
		if open := strings.LastIndexByte(text, '('); open >= 0 {
			note := text[open+1 : len(text)-1]
			if !strings.ContainsAny(note, "()") {
				text = strings.TrimSpace(text[:open])
			}
		}
	}
	stems := make([]string, 0)
	seen := make(map[string]bool)
	for token := range strings.SplitSeq(text, ",") {
		token = strings.TrimSpace(token)
		switch strings.ToLower(token) {
		case "", "-", "none", "n/a":
			continue
		}
		if !validSpecStem(token) {
			return nil, false
		}
		if !seen[token] {
			seen[token] = true
			stems = append(stems, token)
		}
	}
	return stems, true
}
func validSpecStem(stem string) bool {
	for index, character := range []byte(stem) {
		if character >= 'A' && character <= 'Z' ||
			character >= 'a' && character <= 'z' ||
			character >= '0' && character <= '9' ||
			index != 0 && (character == '.' || character == '_' || character == '-') {
			continue
		}
		return false
	}
	return stem != ""
}

// AddedSpecEvidence returns spec stems named by rows added or rewritten in the
// selected worktree shards. Byte-level table reformatting does not count: rows
// are paired by their five canonical cells against HEAD before new stems are read.
func AddedSpecEvidence(tree string, paths []string) ([]string, []string, error) {
	headPaths, err := classFilesAtHead(tree)
	if err != nil {
		return nil, nil, err
	}
	atHead := make(map[string]bool, len(headPaths))
	for _, path := range headPaths {
		atHead[path] = true
	}
	stems := make([]string, 0)
	seen := make(map[string]bool)
	malformed := make([]string, 0)
	for _, path := range paths {
		current, err := os.ReadFile(filepath.Join(tree, filepath.FromSlash(path))) //nolint:gosec // the path is a journal shard under the checkout root
		if err != nil {
			return nil, nil, fmt.Errorf("read journal shard %s: %w", path, err)
		}
		currentRows, valid := evidenceRows(string(current))
		if !valid {
			malformed = append(malformed, path)
			continue
		}
		headRows := []journalRow(nil)
		if atHead[path] {
			answer := runGit(tree, "show", headObject(path))
			if answer.startErr != nil || answer.code != 0 {
				return nil, nil, gitReadError("read "+headObject(path), answer)
			}
			headRows, valid = evidenceRows(string(answer.stdout))
			if !valid {
				malformed = append(malformed, path)
				continue
			}
		}
		held := make(map[[5]string]int, len(headRows))
		for _, row := range headRows {
			held[row.cells]++
		}
		for _, row := range currentRows {
			if held[row.cells] != 0 {
				held[row.cells]--
				continue
			}
			rowStems, _ := specStems(row.cells[1])
			for _, stem := range rowStems {
				if !seen[stem] {
					seen[stem] = true
					stems = append(stems, stem)
				}
			}
		}
	}
	sort.Strings(malformed)
	return stems, malformed, nil
}

func evidenceRows(contents string) ([]journalRow, bool) {
	rows := journalRows(contents)
	for _, row := range rows {
		if row.malformed {
			return nil, false
		}
		if _, valid := specStems(row.cells[1]); !valid {
			return nil, false
		}
	}
	return rows, true
}

// HeadSpecEvidence returns the committed journal rows that name a spec stem.
// The map value is the journal shard carrying the evidence. A malformed shard
// contributes no evidence and is named separately so callers fail closed.
func HeadSpecEvidence(tree string) (map[string]string, []string, error) {
	paths, err := classFilesAtHead(tree)
	if err != nil {
		return nil, nil, err
	}
	classes, err := readClasses(tree, paths)
	if err != nil {
		return nil, nil, err
	}
	evidence := make(map[string]string)
	malformed := make([]string, 0)
	for _, class := range classes {
		path := classPath(class.name)
		valid := true
		for _, row := range class.rows {
			if row.malformed {
				valid = false
				break
			}
		}
		if !valid {
			malformed = append(malformed, path)
			continue
		}
		classEvidence := make([]string, 0)
		for _, row := range class.rows {
			rowStems, readable := specStems(row.cells[1])
			if !readable {
				valid = false
				break
			}
			classEvidence = append(classEvidence, rowStems...)
		}
		if !valid {
			malformed = append(malformed, path)
			continue
		}
		for _, stem := range classEvidence {
			if _, exists := evidence[stem]; !exists {
				evidence[stem] = path
			}
		}
	}
	sort.Strings(malformed)
	return evidence, malformed, nil
}
